File: sorting/bubble_sort.py
"""
Passes over the list, 
catches the maximum/minimum element, 
and brings it over to the right side.
"""


def bubble_sort(lst):
    for i in range(len(lst)):
        for j in range(len(lst) - i, -1):
            if lst[j] > lst[j + 1]:
                lst[j], lst[j + 1] = lst[j + 1], lst[j]


def sort_binary_list(lst):
    # filter each item and compare them with the prev
    # if item is less than prev, swap them
    j = 0
    for i in range(len(lst)):
        if lst[i] < 1:
            lst[i], lst[j] = lst[j], lst[i]
            j += 1

    return lst


def find_max_prod(lst):
    """
    Finds the pair having maximum product in a given list
    :param lst: A list of integers
    :return: A pair of integer
    """
    sorted_lst = sorted(lst)
    return sorted_lst[-2], sorted_lst[-1]

File: sorting/bubbleSort.cpp
#include <vector>
#include <algorithm>

/*
Passes over the array,
catches the maximum/minimum element,
and brings it over to the right side.
*/

void bubbleSort(std::vector<int>& arr)
{
    for (int i = 0; i < arr.size(); i++)
    {
        for (int j = arr.size() - i; j > 0; j--)
        {
            if (arr[j] > arr[j + 1])
                arr[j], arr[j + 1] = arr[j + 1], arr[j];
        }
    }
}

std::vector<int> sortBinaryArray(std::vector<int>& arr)
{
    // filter each item and compare them with the prev
    // if item is less than prev, swap them
    int j = 0;
    for (int i = 0; i < arr.size(); i++)
    {
        if (arr[i] < 1)
            arr[i], arr[j] = arr[j], arr[i];
        j++;
    }
    return arr;
}

std::pair<int, int> findMaxProd(std::vector<int>& arr)
{
    /*
    Finds the pair having maximum product in a given array
    :param arr: A array of integers
    :return: A pair of integers
    */

    std::sort(arr.begin(), arr.end());
    std::pair<int, int> sortedPair = { arr[-2], arr[-1] };

    return sortedPair;
}